Global refactoring

vrslev opened the pull request:
# Summary

What was done:
- Clean up, unify and _uncomplicate_ the code: some things were removed, some merged together.
- Refactor integration tests: now they're written in Python, more efficient, precise and meticulous 😃
- Refactor DevOps utils: much faster, less flaky, easier to maintain and use.

You can check out changes yourself locally. To do this, clone [my fork](https://github.com/vrslev/frappe_docker/tree/beautify), checkout to `beautify` branch. Then build images with `FRAPPE_VERSION=... ERPNEXT_VERSION=... docker buildx bake`. Next step is to start everything up, here are docs: https://github.com/vrslev/frappe_docker/blob/beautify/docs/images-and-compose-files.md, https://github.com/vrslev/frappe_docker/blob/beautify/docs/setup-options.md.

After the PR is merged, your current setup _will_ break. There're no migration guide, but we (@revant and I) updated [the docs](https://github.com/vrslev/frappe_docker/blob/beautify/README.md), so it shouldn't be a problem. If you experience difficulties with upgrading, feel free to leave an issue.

---
# Why?

Currently there's a lot of stuff that could be done better. I'm trying to improve performance, logic behind processes and maintainer happiness.

# What?

Almost everything.

## CI

- Move `get-latest-tags.sh` to Python (`get_latest_tags.py`). Kinda preparation for #616.
- Add `docker-build-push.yml` reusable workflow and use it in `build_develop.yml` and `build_stable.yml`. It is clearer this way.
- Remove `test.yml` and add integration test to build workflows.
- Correct `dependabot.yml` according to changes.

## Images

- Rename target names in frappe/bench.
- Remove frappe/erpnext-worker and frappe/erpnext-nginx separate Dockerfiles. They are now embedded in frappe/frappe-worker and frappe/frappe-nginx Dockerfiles using multi-stage builds. It is much more efficient this way.
- Move nginx Dockerfile from frappe-nginx to nginx and worker Dockerfile from frappe-worker to worker.
- Don't include development node modules in final nginx images.
- Reuse stages for efficient builds in nginx images.
- Update Nginx config from frappe/bench upstream.
- Make Nginx config more opinionated. Who wants can mount their own.
- Simplify frappe/frappe-socketio build.
- Remove majority of frappe/frappe-worker commands since there's pretend bench (auto_migrate, backup, check_connection, constants, doctor, gevent_patch, migrate, new, restore_backup and utils modules are gone).
- Refactor pretend bench (frappe's `bench_helper.py` wrapper).
- Patch database initialization (in pretend bench) to allow any host for db user. Previously it was done in `new` command in `docker-entrypoint.sh`
- Remove `docker-entrypoint.sh` from frappe/frappe-worker (using `configurator` service in compose for config generation)
- Refactor `push_backup.py` in frappe/frappe-worker.
- Reuse stages for efficient builds in worker images.
- Add `configure.py` for dynamic `common_site_config.json` generation. Previously user had to remove that file after updating values in compose file that has affect on networking.

## Bake file

- Remove redundant targets. It is just better now.

## Compose files

I changed structure of overrides, renamed `installation`  directory to `overrides`, renamed `docker-compose.*.yml` to `compose.*.yaml`. This follows compose-spec and allows to use `docker buildx bake` without specifying bake file.
Now there's one main `compose.yaml` and three overrides: ERPNext, HTTPS and Postgres.

About main compose file:
- Add `configurator` service that generates `common_site_config.json` on every `docker-compose up`
- Remove some unnecessary volume mappings.
- Simplify service names, e.g. `frappe-python` to `backend`. Also using unified service names helps with overrides.
- Use YAML merge types: https://yaml.org/type/merge.html to reduce code size.
- More opinionated environment variables setup. User doesn't need to specify variables like `MARIADB_HOST`.
- Remove `env-example`, `env-local`, `env-production`, create unified `example.env` and add useful comments.

## Tests

Now Python is used for integration testing. It is much more pleasant to write, read and debug.
I took advantage of compose overrides for testing in CI. They replace images with the ones in local registry.
I added test for nginx configuration (whether user will be able to access endpoints from different containers or not).
Now there's no issues while testing locally.

## Docs

- Remove Play With Docker config and readme note for now.
- Rewrite the readme.
- Add `custom_app` directory with custom app guide and baseline of `docker-bake.hcl`, `compose.override.yml` and Dockerfiles.
- Add quick guide for Docker images, compose files and environment variables.
- Rewrite site operations doc.
- Remove tips for moving deployment doc.

## Devcontainer
- Move default settings inside container to from `settings.json` to `devcontainer.json`.
- Style `docker-compose.yml` a little.
- Fix inconsistencies in readme.

## Other

- Fix `.dockerignore`.
- Fix `.gitignore`.
- Add `check-executables-have-shebangs` and `check-shebang-scripts-are-executable` hooks in pre-commit config.
- Remove `.prettierignore`, don't need it anymore.
- Add Python requirements for development.

---

**To do before merge**
- [x] Move some things from wiki to /docs and remove it
- [x] Remove/change docker swarm guide
- [x] Modify single-bench guide to single-server; decide what to do with multi-bench guide
- [x] Debug update example env job in build_stable.yml
- [x] Test with helm chart
- [ ] Fix docker-build-push workflow path from vrslev/frappe_docker@beautify to frappe/frappe_docker@main


**To do after merge**
- [ ] Rewrite CONTRIBUTING.md
- [ ] Bring Play With Docker back
- [x] Add [inline cache](https://github.com/moby/buildkit/blob/master/frontend/dockerfile/docs/syntax.md#run---mounttypecache) in Dockerfiles. This will speed up local builds and experiments (like this one).

Closes #606.

vrslev commented:
@revant Can you review and help with todos before merge?

revant commented on build/worker/Dockerfile:
```diff
@@ -0,0 +1,116 @@
+ARG FRAPPE_VERSION
+RUN git clone --depth 1 -b ${FRAPPE_VERSION} https://github.com/frappe/frappe apps/frappe \
+    && env/bin/pip install --no-cache-dir -e apps/frappe \
+    && rm -r apps/frappe/.git
```
Sometimes it's helpful to identify or verify the latest commit hash and know if the container is running the required code.

vrslev replied:
It is a good practice to remove `.git` folder (usually it is done in `.dockerignore` in case if Dockerfile located in main repo. If something changes in code, version of an app should change, too.

revant replied:
ideally yes version should change.

Frappe bump doesn't cause change in ERPNext version

revant commented on example.env:
```diff
@@ -0,0 +1,17 @@
+# or if host is `127.0.0.1` (local debugging), it should be `127.0.0.1` This variable allows
+# to override described behavior. Let's say you create site named `mysite`
+# and do want to access it by `127.0.0.1` host. Than you would set this variable to `mysite`.
+FRAPPE_SITE_NAME_HEADER=
```
This restricts the bench to one site?

vrslev replied:
Only if the variable set. I saved current behavior. Default value of this variable set in `compose.yaml` to `$$host` (escaped from `$host`).

revant commented on example.env:
```diff
@@ -0,0 +1,17 @@
+# Reference: https://github.com/frappe/frappe_docker/blob/main/docs/images-and-compose-files.md
```
do any environment variables change?

they are used by containers here https://github.com/frappe/helm/tree/main/erpnext/templates in deployment-*.yaml

vrslev replied:
They did. All of them are in [`example.env`](https://github.com/vrslev/frappe_docker/blob/beautify/example.env). Also, [docs](https://github.com/vrslev/frappe_docker/blob/beautify/docs/images-and-compose-files.md#configuration).

revant replied:
is backward compatibility possible? I'll have to heavily change charts.

vrslev replied:
It is already partially backwards compatible. Only variable name that changed is `DB_PASSWORD` (previously, it was `MYSQL_ROOT_PASSWORD` and `POSTGRES_PASSWORD`). It had to be done for Postgres override.

Other variables (`ADMIN_PASSWORD`, `MARIADB_HOST`, `ENTRYPOINT_LABEL` and similar) just not used anymore.

revant commented on tests/_check_backup_files.py:
```diff
@@ -0,0 +1,43 @@
+        raise exc("private files")
+    if not public_files:
+        raise exc("public files")
+    print("All files was pushed to S3!")
```
"All files **were** pushed to S3" ??

vrslev replied:
```suggestion
    print("All files were pushed to S3!")
```

revant commented on tests/_create_bucket.py:
```diff
@@ -0,0 +1,19 @@
+        service_name="s3",
+        endpoint_url="http://minio:9000",
+        region_name="us-east-1",
+        aws_access_key_id=os.getenv("MINIO_ACCESS_KEY"),
```
Tests just use MINIO, but why env var to be called MINIO_, It's actually `S3_ACCESS_KEY` and `S3_SECRET_KEY`

vrslev replied:
I took this variables from current tests, but sure: https://github.com/frappe/frappe_docker/blob/c2b15431281e470678676bd6b1d9ac0288536850/tests/integration-test.sh#L141-L142 Fixed: https://github.com/vrslev/frappe_docker/commit/5e16ed89b6695eab7b486164c2e13b5cdf750de8

revant commented on tests/testfile.txt:
```diff
@@ -0,0 +1 @@
+lalala
```
What's this?

vrslev replied:
This file used to test if `/files` endpoint works: https://github.com/vrslev/frappe_docker/blob/5e16ed89b6695eab7b486164c2e13b5cdf750de8/tests/main.py#L223

vrslev commented on .github/workflows/build_develop.yml:
```diff
@@ -33,82 +31,32 @@ jobs:
+
+  build_main:
+    name: Frappe and ERPNext
+    uses: vrslev/frappe_docker/.github/workflows/docker-build-push.yml@beautify
```
This is required for now. Before merging I will change it to main repo and branch. In before merge todos: `Fix docker-build-push workflow path from vrslev/frappe_docker@beautify to frappe/frappe_docker@main`

vrslev commented on build/nginx/Dockerfile:
```diff
@@ -0,0 +1,88 @@
+COPY --from=frappe_prod_node_modules /root/frappe-bench/apps/frappe/node_modules /usr/share/nginx/html/assets/frappe/node_modules
+COPY --from=frappe_assets /root/frappe-bench/sites /usr/share/nginx/html
+
+COPY nginx-template.conf /
```
Yep. See https://github.com/vrslev/frappe_docker/blob/5e84d7734c362d86e95b05dc3da72b548082f3c2/build/nginx/Dockerfile#L80
Also, it it okay that container uses `nginx` image instead of `nginxinc/nginx-unprivileged`?

revant replied:
not really,

if you try to make normal nginx image run as non root then it won't work. it needs additional configuration.

not good to run apps in non root containers.

vrslev replied:
> not good to run apps in non root containers.

It is good to run app in non-root containers — less privileges.

So, I should use `nginxinc/nginx-unprivileged`, right?

revant replied:
Yes like before, with following

https://github.com/frappe/frappe_docker/blob/c2b15431281e470678676bd6b1d9ac0288536850/build/frappe-nginx/Dockerfile#L52-L52

vrslev replied:
I updated compose file, nginx template and Dockerfile. I used `1101` user instead of `nginx` and it works without changes use suggested:
https://github.com/vrslev/frappe_docker/blob/225c69709780e7dd2e282aafd9e16e63654d3ed1/build/nginx/Dockerfile#L71-L90

revant replied:
https://github.com/frappe/frappe_docker/pull/307 this needed the users to match.
It was for issue https://github.com/frappe/frappe_docker/pull/299

I don't think it exists anymore. Can you confirm it didn't introduce back 299

vrslev replied:
I will later check if cache works without changes.

vrslev replied:
It did not work without changes! It works now: https://github.com/frappe/frappe_docker/pull/617/commits/075df53a337e234ad297b8e84ef583be3ab7f25e. Can you review this?

ddavidebor reviewed (approved):
Looks like amazing work

revant commented:
> Can you help me with docs?

sure, commit on the same branch?

> I don't think we should keep Docker Swarm docs. It is much more advanced than plain docker compose. If someone would want to use it, they'd have to learn it carefully, and this kind of guide is not enough.

Maybe simple `docker stack deploy -c compose.yml erpnext` should be there. That will also enable PWD.

> Also, I want to merge single bench and multiple bench guides, and rewrite it all.

Most generally setup 1 bench and add multiple sites to it. Some may use multiple benches to host different branches of apps.
For single bench traefik/nginx-proxy is part of the docker-compose. For multiple benches we need the traefik/nginx-proxy to be separate from the benches.

> Moreover, I would like to transfer all guides from wiki to /docs so we'd have single source of truth.

Okay, it's not costly to add docs any more. Merging docs PR won't trigger unnecessary builds.

Most of the things you've done already. I'll come up a checklist of things I'll do.

vrslev commented:
> sure, commit on the same branch?

Yeah, sure.

> Maybe simple docker stack deploy -c compose.yml erpnext should be there. That will also enable PWD.

Maybe it is. I will test it later.

> Most of the things you've done already. I'll come up a checklist of things I'll do.

Happy to cooperate!

vrslev commented:
I was stuck with _generating_ compose file for PWD (and Docker Swarm in general).

PWD requires a single file. It means environment variables and site creator service should be embedded in compose file. But output format of `docker compose config` is incompatible with `docker stack deploy`. The point of refactoring compose file configurations is to have single source of truth. This way users can easier understand what’s going on and what they should do, and maintainers — maintain.

Do you have suggestions for a nice way to generate PWD compose file?

revant commented:
`docker-compose` is not `docker stack`. For docker stack to work docker swarm should be initialized (`docker swarm init`).

Let me check if we can hack in the output of current configs + some `sed` magic to get some stacks for swarm.

revant commented on example.env:
```diff
@@ -0,0 +1,17 @@
+# Reference: https://github.com/frappe/frappe_docker/blob/main/docs/images-and-compose-files.md
+
+FRAPPE_VERSION=v13.16.0
```
How do we bump this with every release?

vrslev replied:
We should. I create a script that would do that.

vrslev replied:
Done: https://github.com/frappe/frappe_docker/pull/617/commits/adc526d8598120977d5d31874688e4057edc0b9c.

revant reviewed (commented):
env file tags version to current stable.

vrslev commented:
> docker-compose is not docker stack. For docker stack to work docker swarm should be initialized (docker swarm init).

Yes, I know that 😃
That was a common practice to do `docker-compose -f docker-compose.yml -f docker-compose.override.yml config >stack.yml && docker stack deploy -c stack.yml`. This way developers have single compose file for deployment and for development.

revant commented:
I was able to do some workarounds to generate stack.yaml

- I removed proxy from main compose.yaml, moved it to `overrides/compose.proxy.yaml`. By this swarm stack can use globally installed traefik. Can be overriden to generate traefik for swarm mode.
- I added compose.swarm.yaml:

```yaml
services:
  frontend:
    deploy:
      restart_policy:
        condition: on-failure
      labels:
        traefik.enable: true
        traefik.http.services.frontend.loadbalancer.server.port: 8080
        traefik.http.routers.frontend-http.entrypoints: websecure
        traefik.http.routers.frontend-http.tls.certresolver: main-resolver
```

I've had clean container labels and depends_on,

```shell
docker-compose -f compose.yaml -f overrides/compose.erpnext.yaml -f overrides/compose.swarm.yaml config | yq eval 'del(.services.*.depends_on) | del(.services.frontend.labels)' -
```

Output:

<details>

```yaml
services:
  backend:
    image: frappe/erpnext-worker:v13.16.1
    volumes:
      - sites:/home/frappe/frappe-bench/sites:rw
      - assets:/home/frappe/frappe-bench/sites/assets:rw
  configurator:
    command: configure.py
    environment:
      DB_HOST: db
      DB_PORT: 3306
      REDIS_CACHE: redis:6379/0
      REDIS_QUEUE: redis:6379/1
      REDIS_SOCKETIO: redis:6379/2
      SOCKETIO_PORT: 9000
    image: frappe/erpnext-worker:v13.16.1
    volumes:
      - sites:/home/frappe/frappe-bench/sites:rw
  db:
    command:
      - --character-set-server=utf8mb4
      - --collation-server=utf8mb4_unicode_ci
      - --skip-character-set-client-handshake
      - --skip-innodb-read-only-compressed
    environment:
      MYSQL_ROOT_PASSWORD: '123'
    healthcheck:
      interval: 1s
      retries: 15
      test: mysqladmin ping -h localhost --password=123
    image: mariadb:10.6
    volumes:
      - db-data:/var/lib/mysql:rw
  frontend:
    deploy:
      labels:
        traefik.enable: "True"
        traefik.http.routers.frontend-http.entrypoints: websecure
        traefik.http.routers.frontend-http.tls.certresolver: main-resolver
        traefik.http.services.frontend.loadbalancer.server.port: '8080'
      restart_policy:
        condition: on-failure
    environment:
      BACKEND: backend:8000
      FRAPPE_SITE_NAME_HEADER: $$$$host
      SOCKETIO: websocket:9000
    image: frappe/erpnext-nginx:v13.16.1
    volumes:
      - assets:/usr/share/nginx/html/assets:rw
      - sites:/usr/share/nginx/html/sites:rw
  queue-default:
    command: bench worker --queue default
    image: frappe/erpnext-worker:v13.16.1
    volumes:
      - sites:/home/frappe/frappe-bench/sites:rw
  queue-long:
    command: bench worker --queue long
    image: frappe/erpnext-worker:v13.16.1
    volumes:
      - sites:/home/frappe/frappe-bench/sites:rw
  queue-short:
    command: bench worker --queue short
    image: frappe/erpnext-worker:v13.16.1
    volumes:
      - sites:/home/frappe/frappe-bench/sites:rw
  redis:
    image: redis:6.2-alpine
    volumes:
      - redis-data:/data:rw
  scheduler:
    command: bench schedule
    image: frappe/erpnext-worker:v13.16.1
    volumes:
      - sites:/home/frappe/frappe-bench/sites:rw
  websocket:
    image: frappe/frappe-socketio:v13.16.0
    volumes:
      - sites:/home/frappe/frappe-bench/sites:rw
version: '3.9'
volumes:
  assets: {}
  db-data: {}
  redis-data: {}
  sites: {}
```

</details>

vrslev commented:
@revant Is this okay? https://github.com/frappe/frappe_docker/pull/617/commits/e23fe73353df274e153e7ca4e51d821122c545f9

revant commented:
> @revant Is this okay? [e23fe73](https://github.com/frappe/frappe_docker/commit/e23fe73353df274e153e7ca4e51d821122c545f9)

did you confirm the issue due to `.build` https://github.com/frappe/frappe_docker/pull/617#discussion_r772861603

vrslev commented:
@revant

> I removed proxy from main compose.yaml, moved it to overrides/compose.proxy.yaml.

I guess, it is fine. We should change docs then.

> I added compose.swarm.yaml

We also have to set `restart_policy` so backend, queues, schedule and websocket services work properly.

vrslev commented:
> did you confirm the issue due to .build #617 (comment)

I don't understand what you're trying to say. Commit I asked you about is irrelevant to cache issue. I ran tests and everything works fine with this change.

revant commented:
> We also have to set `restart_policy` so backend, queues, schedule and websocket services work properly.

`restart_policy` is under `deploy`, I think docker-compose will use `restart` under service. https://docs.docker.com/compose/compose-file/compose-file-v3/#restart

vrslev commented:
> restart_policy is under deploy, I think docker-compose will use restart under service.

I know that. Docker Compose works fine without setting `restart` because app services depend on successful exit of `configurator ` service. But `depends_on` is not used in your implementation for Docker Swarm, that's why we should set `restart_policy`.

revant commented:
> > did you confirm the issue due to .build #617 (comment)
> 
> I don't understand what you're trying to say. Commit I asked you about is irrelevant to cache issue. I ran tests and everything works fine with this change.

cache ro works.

not related directly, the .build file needs to exist for v12 desk

https://github.com/frappe/frappe/blob/cac7cc402fa8e28120284ae49a6cff0fb8681c38/frappe/www/desk.py#L87

vrslev commented:
> cache ro works.

I tested before adding code relevant to `.build` and it didn't work for me. After making some changes it works.

revant commented:
adding .build to nginx image won't be available in python image when it does the mentioned function call

vrslev commented:
> adding .build to nginx image won't be available in python image when it does the mentioned function call

I think we don't understand each other on this one. I repeated previous behavior of `.build` file creation.

revant commented:
> > adding .build to nginx image won't be available in python image when it does the mentioned function call
> 
> I think we don't understand each other on this one. I repeated previous behavior of `.build` file creation.

if the .build file is accessed by the python code?
then whenever it tries to get the attributes of the .build file It'll logically search the `sites` directory mounted in python container.

It won't be able to access the .build file in nginx image?

I think previous behaviour was in entrypoint.sh file, which made sure the file is created in mounted volume instead of image itself.

vrslev commented:
> I think previous behaviour was in entrypoint.sh file, which made sure the file is created in mounted volume instead of image itself.

It is in entrypoint script!
https://github.com/frappe/frappe_docker/blob/23b0ff6454590e8fd97e6c018eced328f16480b4/build/nginx/entrypoint.sh#L1-L9

revant commented:
> It is in entrypoint script!

🙏 Yes it is! 😅

revant commented:
@vrslev can you go through this https://github.com/vrslev/frappe_docker/compare/beautify...revant:beautify-docs?expand=1

revant commented:
Thoughts for custom app building:

- In custom_app repo's ci clone this repo and run script in the repo
- clone all apps needed to be installed in the image during ci in directory called `repos`
  - cloning during ci allows private apps to be cloned.
- Dockerfile(s) copy `repos` to `~/frappe-bench/apps` and install

vrslev commented:
I get what you're trying to do here, but I think it doesn't belong in this repository 😅

vrslev commented:
Also, some parts of building process are implicit and some are not. For example, [in this Dockerfile](https://gitlab.com/gwptcorg/GWPTCRegistrarRepo/-/blob/main/containers/worker/Dockerfile) all apps are being copied from `repos` directory, but only `whitelabel` is installed.

Also, it is better to use `ONBUILD` instruction: https://docs.docker.com/engine/reference/builder/#onbuild

vrslev commented:
> @vrslev can you go through this https://github.com/vrslev/frappe_docker/compare/beautify...revant:beautify-docs?expand=1

I created PR: https://github.com/vrslev/frappe_docker/pull/9

revant commented:
> I get what you're trying to do here, but I think it doesn't belong in this repository sweat_smile

Yes, it doesn't belong here.

But current custom_app docs is also not sufficient for combination of private and public apps. That will be the first question on forum/issues. How to install multiple apps? How to install private apps?

Copying the code from repository makes it obsolete. Cloning or using the repo itself will not become obsolete.

Instead of duplicating files, I use frappe/frappe_docker to build images with [kaniko](https://github.com/GoogleContainerTools/kaniko) here: 

https://github.com/revant/erpnext_feature_board/blob/8c99b4dfaa79d86d8e8b46fa1bf235d0bfa471e0/erpnext_feature_board/clients/k8s.py#L63-L79

vrslev commented:
> How to install private apps?

_How_ one would install app from private repository? By [mounting secrets](https://docs.docker.com/develop/develop-images/build_enhancements/#new-docker-build-secret-information)?

revant commented:
Just a note, I depend on git even in containers, It is useful to check last commit hash in the container and in the app repo and confirm if the changes are in container for custom `latest` tags

revant commented:
v14 beta released.

we want the major refactor to docker, helm and v14 now? this needs to be closed or completed!

vrslev commented:
@revant Yep, I’m waiting for you with PR in my repo

revant commented:
I pushed few changes. please review.

vrslev commented:
@revant Hey there! I will test your changes — I mean, run [and maybe fix] integration test — tomorrow.

Changes I propose are huge. Do you think you will be able to properly maintain them by yourself? I don't mind answering to an issue, helping to resolve one or making a patch. But currently I have a lot in life and don't think I will be able (and to be honest, interested 😅) in being a "maintainer".
I know you do commerce development and DevOps for companies that use Frappe and ERPNext. If new stuff _will_ be a burden for you, I think we should revisit concept of huge PR and make some major smaller improvements.

revant commented:
> I think we should revisit concept of huge PR and make several major but small improvements.

after looking at the complexity of this PR, that'll be great.

> new stuff _will_ be a burden for you

unfortunately that happens everywhere! we've to keep updated with technology!

vrslev commented:
> after looking at the complexity of this PR, that'll be great.
I’ll look into how to do this the best way.

vrslev commented:
@revant You know what: I don't think it would be possible to easily split changes in several PRs. If I want to change Bake file, Dockerfiles have to be changed. If I change Dockerfiles, I have to change helper scripts too. Then I change compose files, because they depend on helper scripts. After all of that, tests don't work!

vrslev commented:
@revant Too much work required to split changes in separate PRs. Can we merge this one and then deal with consequences later? For sure, we have to test the changes with Helm chart.

vrslev commented:
@revant I merged your PR, everything is fine. Also adapted this to new test, now everything works. Can you apply the changes to Helm yourself?

revant commented:
> Can you apply the changes to Helm yourself?

yes I'll check that out and create a branch on helm chart repo

revant commented:
I started work on helm. @vrslev check 2 PRs to your repo

also i thing `configure.py` can be removed. There is `bench  set-config` https://frappeframework.com/docs/v13/user/en/bench/reference/set-config

vrslev commented:
> also i thing configure.py can be removed. There is bench set-config
I gave up on using `bench set-config` — it works very slow

revant commented:
> > also i thing configure.py can be removed. There is bench set-config
> > I gave up on using `bench set-config` — it works very slow

This worked for me.

```
...
      containers:
      - name: configure
        image: "{{ .Values.worker.image.repository }}:{{ .Values.worker.image.tag }}"
        imagePullPolicy: {{ .Values.worker.image.pullPolicy }}
        command: ['bash', '-c']
        args:
          - >
            [[ -f common_site_config.json ]] || echo "{}" > common_site_config.json;
            [[ -f apps.txt ]] || ls -1 ../apps > apps.txt;
            bench set-config -g db_port $DB_PORT;
            bench set-config -g db_host $DB_HOST;
            bench set-config -g redis_cache $REDIS_CACHE;
            bench set-config -g redis_queue $REDIS_QUEUE;
            bench set-config -g redis_socketio $REDIS_SOCKETIO;
            bench set-config -g socketio_port $SOCKETIO_PORT;
...
```
apps.txt was empty when new volume for sites was mounted


gevent from frappe-worker python env was removed?

vrslev commented:
> gevent from frappe-worker python env was removed?

I don’t think it is necessary in _official_ image

revant commented:
> I don’t think it is necessary in _official_ image

It's being used now with env variables. it'll break compatibility if anyone using it.

jq is kept other tools are there in image that are not officially used by frappe. just adding 1 pip dependency that may increase io performance for some cases won't be a problem.

I'll add with a PR

vrslev commented:
@revant Okay. Let me know when you will be mostly done 😉

revant commented:
I've a branch for beautify for my fork of helm.
I'm fixing --mariadb-root-* to --db-root-* on Frappe/Frappe.

revant commented:
@vrslev review changes on your branch

vrslev commented:
> I'm fixing --mariadb-root-* to --db-root-* on Frappe/Frappe.
@revant Do we need to wait?

vrslev commented:
Stable build is supposed to fail until PR is merged: https://github.com/frappe/frappe/pull/16035

vrslev commented:
@revant Are we ready to finally merge this PR?

revant commented:
> @revant Are we ready to finally merge this PR?

no! Let the PR take it's time! Even if it is lost it doesn't matter.

revant commented:
Two things remaining for Helm Charts

- [ ] Documentation
- [ ] Test helm migration from v3.x.x to v4.x.x locally
- [ ] Automated tests for testing PRs

revant commented:
> @revant Do we need to wait?

No. We proceed with refactor of frappe_docker and helm.

vrslev commented:
> Things remaining for Helm Charts

So we wait? 😃

revant commented:
do you have any suggestion for generating changelog?

for this major thing and then onwards.

vrslev commented:
I don’t think generating changelog automatically is a good idea. 

Previously I used [python-semantic-release](https://github.com/relekang/python-semantic-release) in my projects for automatic releases ([example](https://github.com/vrslev/ikea-api-client/commit/bf005998c603b12bbedb0735ce1f4a4db14b9f26)). It bumps version in code, updates changelog with commit messages based on semantic commits spec (“feat”, “fix” etc), creates GitHub release.

Now I don’t see any value in keeping CHANGELOG.md when there’s GitHub Releases. I create releases and write release notes manually ([example](https://github.com/vrslev/ikea-api-client/releases/tag/v2.0.3)) but use [this GitHub feature](https://docs.github.com/en/repositories/releasing-projects-on-github/automatically-generated-release-notes) when can. 

Any other automatic release management tools ([Release Drafter](https://github.com/release-drafter/release-drafter), [Semantic Release](https://github.com/semantic-release/semantic-release)) assume that you also manage version of software. How would you do that with frappe_docker?

revant commented:
For Changelog I think manual entry seems right. May be ask PRs to add their list of features in a file?
In that case there is no release as such. Only list of changes on `main` branch.

`push-backup` command is not working as expected. I tried with my credentials, it creates `/` and `home` directory inside `/` in the bucket. I couldn't find the pushed backups.

command:

```shell
bench --site=$(SITE_NAME) backup --with-files; push-backup --site=$(SITE_NAME) --bucket=$(BUCKET) --region-name=$(REGION) --endpoint-url=$(ENDPOINT_URL) --aws-access-key-id=$(ACCESS_KEY) --aws-secret-access-key=$(SECRET_KEY);
```

Output:

```
Backup Summary for erp.cluster.local at 2022-02-22 15:43:44.068947
Config  : ./erp.cluster.local/private/backups/20220222_154343-erp_cluster_local-site_config_backup.json 147.0B
Database: ./erp.cluster.local/private/backups/20220222_154343-erp_cluster_local-database.sql.gz         227.5KiB
Public  : ./erp.cluster.local/private/backups/20220222_154343-erp_cluster_local-files.tar               10.0KiB
Private : ./erp.cluster.local/private/backups/20220222_154343-erp_cluster_local-private-files.tar       10.0KiB
Backup for Site erp.cluster.local has been successfully completed with files
['--site=erp.cluster.local', '--bucket=xxx', '--region-name=nl-ams', '--endpoint-url=http://s3.nl-ams.scw.cloud', '--aws-access-key-id=EXAMPLE', '--aws-secret-access-key=3e080506-dc57-4701-8fe3-0539859bb043']
Uploading files: ['./erp.cluster.local/private/backups/20220222_154343-erp_cluster_local-database.sql.gz', './erp.cluster.local/private/backups/20220222_154343-erp_cluster_local-files.tar', './erp.cluster.local/private/backups/20220222_154343-erp_cluster_local-private-files.tar', './erp.cluster.local/private/backups/20220222_154343-erp_cluster_local-site_config_backup.json']
```

vrslev commented:
> Only list of changes on main branch.

Good idea.

>  it creates / and home directory inside / in the bucket

I don't understand 😅 Can you share an example how it works now and how is supposed to work?

vrslev commented:
@revant Any news?

- Remaining tasks regarding Helm chart — do we have to wait for the job done or do we go ahead and merge this PR?
- Broken push-backup script — can you show on practical example what goes wrong?

revant commented:
> * Remaining tasks regarding Helm chart — do we have to wait for the job done or do we go ahead and merge this PR?

I've everything working for new image in new chart with all of the documentation except the part of chart migration.

things remaining:

1) update automated tests
2) I've to test and write migration steps for current users to move to new version.

> * Broken push-backup script — can you show on practical example what goes wrong?

```
Backup Summary for erp.cluster.local at 2022-02-25 12:29:27.864941
Config  : ./erp.cluster.local/private/backups/20220225_122927-erp_cluster_local-site_config_backup.json 147.0B
Database: ./erp.cluster.local/private/backups/20220225_122927-erp_cluster_local-database.sql.gz         227.7KiB
Public  : ./erp.cluster.local/private/backups/20220225_122927-erp_cluster_local-files.tar               10.0KiB
Private : ./erp.cluster.local/private/backups/20220225_122927-erp_cluster_local-private-files.tar       10.0KiB
Backup for Site erp.cluster.local has been successfully completed with files
['--site=erp.cluster.local', '--bucket=BUCKETNAME', '--region-name=REGION', '--endpoint-url=URL', '--aws-access-key-id=ACCESSKEY', '--aws-secret-access-key=SECRETKEY']
Uploading files: ['./erp.cluster.local/private/backups/20220225_122927-erp_cluster_local-database.sql.gz', './erp.cluster.local/private/backups/20220225_122927-erp_cluster_local-files.tar', './erp.cluster.local/private/backups/20220225_122927-erp_cluster_local-private-files.tar', './erp.cluster.local/private/backups/20220225_122927-erp_cluster_local-site_config_backup.json']
```

It is printing the command here? It shows the secrets in the output

Messed up push on bucket:

I can see directory called `/`

![00-push-backup-root](https://user-images.githubusercontent.com/1040161/155670345-71d1fe3a-b75d-4183-95e7-3ef23440eb35.png)

then `home`
![01-push-backup-home](https://user-images.githubusercontent.com/1040161/155670412-27a31f4d-0aaa-4ce3-9f37-d6c50014d88c.png)

then it's empty:
![02-push-backup-empty](https://user-images.githubusercontent.com/1040161/155670435-22c88d67-799d-418b-a1ee-d963de969fca.png)

I'll suggest let the prefix path be configurable here (default to just the site name may be or make it mandatory):

https://github.com/vrslev/frappe_docker/blob/cea5f5a9ed275f804c73e0ae116fe70f93ea672c/build/worker/push_backup.py#L58

Another suggestion: the "build" directory which has all the files is ignored by github's find file feature. Can we call it "containers" :grimacing: :sweat_smile:

vrslev commented:
> It is printing the command here? It shows the secrets in the output

I will fix that.

> then it's empty:

Tested with MinIO, it works for me:

```sh
docker compose -p test -f compose.yaml -f overrides/compose.mariadb.yaml -f overrides/compose.redis.yaml -f overrides/compose.noproxy.yaml up -d
docker run --name minio -d -p 9000:9000 -p 9001:9001 --network test_default minio/minio server /data --console-address ":9001"
docker compose -p test exec backend bench new-site test --mariadb-root-password 123 --admin-password admin
docker compose -p test exec backend bench backup --with-files
docker compose -p test cp tests/_create_bucket.py backend:/tmp/
docker compose -p test exec -e S3_ACCESS_KEY=minioadmin -e S3_SECRET_KEY=minioadmin backend ../env/bin/python /tmp/_create_bucket.py
docker compose -p test exec backend push-backup --site test --bucket frappe --region-name us-east-1 --endpoint-url http://minio:9000 --aws-access-key-id minioadmin --aws-secret-access-key minioadmin
```

Then open http://127.0.0.1:9001/buckets/frappe.

> I'll suggest let the prefix path be configurable here (default to just the site name may be or make it mandatory):

I will fix this too.

> Another suggestion: the "build" directory which has all the files is ignored by github's find file feature. Can we call it "containers"

What about "images", "dockerfiles" or "docker"?

vrslev commented:
> I've everything working for new image in new chart with all of the documentation except the part of chart migration.

Do you want to merge this PR along with Helm changes?

revant commented:
> Do you want to merge this PR and Helm changes at the same time?

Yes!

If not then helm charts will start failing due to changes.

I'm done with testing migration. Now only automated tests remain.

vrslev commented:
> Now only automated tests remain.

Great! I guess, the last major step.

> "images" is okay.

Okay, I'll rename the directory. I didn't want it to be "containers" because containers are like _instances_ of images.

> docker is also a brand name. I know repo name is frappe_docker. 

😂

vrslev commented:
Great! Thanks for your work ☺️

revant commented:
make this ready for review?

vrslev commented:
@revant

> can you make a post/note on what changes and what community should be aware of?

What kind of note?

> make this ready for review?

I need to change docker-build-push workflow path from vrslev/frappe_docker@beautify to frappe/frappe_docker@main. It will break builds until PR is merged. Should I do it now?

revant commented:
> What kind of note?

- why refactor? what improved?
- how people can review it, or try it?
- what people can expect on upgrade?
- what will break in existing setup? how to fix?

I'll make it into forum post and pin it for some days as invitation and heads up.

> Should I do it now?

let it be your fork for few days, if someone from community has any suggestion or improvements, let's change it last.

vrslev commented:
@revant I added a summary to the PR message. Can you copy-paste it from there?

revant commented:
I had to workaround to push to my docker registry. USERNAME=revantnandgaonkar didn't work, USERNAME is used from shell instead. which is `revant` locally.

I worked around with add USERNAME to `/tmp/env` file and execute:

```
env $(cat /tmp/env | grep -v "#" | xargs) docker buildx bake --push -f docker-bake.hcl
```

These images can run on existing setups if following changes are made

https://github.com/frappe/frappe_docker/compare/e1f32b747634c392c3dde3390a85dc4cc0505933...compat

vrslev commented:
> USERNAME is used from shell instead

That’s true. `docker build bake` doesn't use .env

revant commented:
https://discuss.erpnext.com/t/2022-q1-refactor-docker-containers-and-helm-chart/87034

vrslev commented:
@revant How long we will wait until merging?

revant commented:
Enjoy the weekend and merge on Monday morning?

Hope that gives some time for forum members to participate in discussion.

The PR is open since December 2021, plenty of time for all to review actually.

vrslev commented:
> Enjoy the weekend and merge on Monday morning?

Cool!

> @vrslev you can do the final touches if you need. Resolve conflicts, anything else you need to change.

Done! Expected failures: https://github.com/frappe/frappe_docker/actions/runs/1956054857, https://github.com/frappe/frappe_docker/actions/runs/1956054856. After merging the PR everything will work fine.

revant commented:
https://github.com/frappe/frappe_docker/issues/606#issuecomment-1061801822

> I still find the current implementation too hard to understand, but it's up to you if you simplify it or not.

@kevgeni what do you find difficult? we can try to simplify or explain better in docs?

revant commented:
@vrslev there was a release few hours ago.

kurtrottmann commented:
Hi, I came from the ERPNext forum, thank you very much for this refactoring, it's an amazing job.

I would like to ask if you can document the correct commands to build the image manually, because is new to me the multi-stage build feature.

What is the correct way to manually build the images? For example, I'm trying to build a socketio image with:

`docker build -t frappe/frappe-socketio:v13.22.0 ./images/socketio --build-arg FRAPPE_VERSION=v13.22.0`

but I get:

```
Step 8/13: RUN mkdir -p sites apps/frappe
   ---> Running on 9354c2df6f83
mkdir: Cannot create directory 'sites': Permission denied
mkdir: Cannot create directory 'applications/': Permission denied
```

I had to edit the Dockerfile and solve the folder problem to succesfully build the image.
 
I built the other images with:

`docker build --target erpnext -t frappe/erpnext-nginx:v13.22.0 ./images/nginx --build-arg FRAPPE_VERSION=v13.22.0 --build-arg ERPNEXT_VERSION=v13.22.0`

`DOCKER_BUILDKIT=1 docker build --target erpnext -t frappe/erpnext-worker:v13.22.0 ./images/worker --build-arg PYTHON_VERSION=3.9 --build-arg FRAPPE_VERSION=v13.22.0 --build-arg ERPNEXT_VERSION=v13.22.0`

to get ERPNext working succesfully, but if I try to create a `frappe-nginx` (Frappe Only) image with:

`docker build --target frappe -t frappe/frappe-nginx:v13.22.0 ./images/nginx --build-arg FRAPPE_VERSION=v13.22.0`

build fails asking for `ERPNEXT_VERSION` variable, which I assume is not necessary. 

**Other comments about the Documentation**:

1. The documentation shows in many parts `docker-compose --project <project-name>`  but must be `docker-compose --project-name <project-name>`.

2. When I run `docker-compose exec backend bench new-site ...` I get the error:
```
service "backend" is not running container #1
```
I have to add project-name parameter to run the command succesfully: `docker-compose --project-name <project-name> exec backend bench new-site ...`

3. It was not obvious to me that I need to manually create a first new site and install ERPNext manually after initialize the containers, because this was created automatically by "frappe_docker" repo previous to refactoring.
_________
Docker version 20.10.12, build e91ed5707e
Docker Compose version 2.3.3

kevgeni commented:
> [#606 (comment)](https://github.com/frappe/frappe_docker/issues/606#issuecomment-1061801822)
> 
> > I still find the current implementation too hard to understand, but it's up to you if you simplify it or not.
> 
> @kevgeni what do you find difficult? we can try to simplify or explain better in docs?

Hello there
That comment was on the issue that was being closed with no changes. But from the look of it, some changes are being accepted/merged in this pull request so all is good. 

I've looked at the beautify branch to be merged and there seems to be quite a lot of changes from the last time I checked. 

If I want to deploy a custom app with docker, how would I do that ? 

I've looked at the latest version. I can't use /custom_app/readme.md, because the use case is a custom app for frappe. 
What I developped is a custom app to be added on top of erpnext. 
From the look of it
- I need to customize the images/nginx and images/worker to add my app in the dockerfile.
  + Last time I could not figure out how to customize the worker. I had customapp_builder, erpnext_builder and frappe_builder and it was overwhelming when you're novice with multi-stage build. I did put time on it, did some tests but it wasn't working and was complex to understand for me and my colleagues. That's where I ended it and took a different route for our containerization. 
- I need to customize the docker-bake.hcl to add new targets and groups

Otherwise, quick questions 
- overrides/compose.https.yaml : what is the url the certificate is requested for ? Does it automatically take the name of the frappe site ? Does not seems to be written in the /docs/images-and-compose-files.md

kevgeni commented:
Quick comment in the beautify branch. 
the docker-bake.hcl uses ${USERNAME} in the function "tag"

On windows, 
docker buildx bake -f docker-bake.hcl frappe-nginx
[+] Building 0.0s (0/0)
error: failed to solve: invalid reference format: repository name must be lowercase

this is caused because my 
echo %USERNAME% envirionment variable contains a upper case character. 
I need to manually change it to something else without uppercase for the build to be able to start.

revant commented:
@kurtrottmann

> because this was created automatically by "frappe_docker" repo previous to refactoring.

The site-creator container had few scripts and waits that did the waiting for setup to be ready and then create site. Do you think it is a blocker for merging?

@kevgeni

> Quick comment in the beautify branch.

I had to do this workaround https://github.com/frappe/frappe_docker/pull/617#issuecomment-1060377681

Changing the variable from USERNAME to CONTAINER_REGISTRY makes any difference?

USERNAME seems to be shell variable everywhere.

kurtrottmann commented:
@revant 

> The site-creator container had few scripts that did the waiting for setup to be ready and then create site. Do you think it is a blocker for merging?

The new procedure is more cleaner and gives more control. My comment was related to mention this extra step more clearly in documentation. I hope you can merge it soon.

Another comment. I noticed that `compose.https.yaml` lacks `- /var/run/docker.sock:/var/run/docker.sock:ro` volume definition which is present in `compose.proxy.yaml`, then I need to use `compose.https.yaml` together with `compose.proxy.yaml` to get access to Frappe/ERPNext via https successfully. However documentation examples says that use  `compose.https.yaml` exclusively is enough to get https working.

vrslev commented:
@kurtrottmann

> 1. Please put the FRAPPE_VERSION=... ERPNEXT_VERSION=... docker buildx bake command explicit in the documentation. This was my first time with docker buildx and it was not easy to understand how to build the images.

@revant Should we update `CONTRIBUTING.md` before merge?

> 2. The documentation shows in many parts docker-compose --project <project-name> but must be docker-compose --project-name <project-name>.

Fixed, thanks!

> 3. When I run docker-compose exec backend bench new-site ... I get the error:

Added note to this doc :)

vrslev commented:
@kevgeni 

> I've looked at the latest version. I can't use /custom_app/readme.md, because the use case is a custom app for frappe.

This is a valid question. One solution is to make custom app example images on top of ERPNext. If somebody would want to adapt setup to non-ERPNext installation, it would be quite easy to strip ERPNext-specific code. @revant, thoughts?

> overrides/compose.https.yaml : what is the url the certificate is requested for ? Does it automatically take the name of the frappe site ? Does not seems to be written in the /docs/images-and-compose-files.md

@revant Could you answer this one? Also we probably should update the docs.

vrslev commented:
@revant

> Changing the variable from USERNAME to CONTAINER_REGISTRY makes any difference? 

Renamed to `REGISTRY_USER`. `CONTAINER_REGISTRY` is not quite correct. Hope you don't mind.

vrslev commented:
@kurtrottmann 

> I noticed that compose.https.yaml lacks - /var/run/docker.sock:/var/run/docker.sock:ro volume definition which is present in compose.proxy.yaml, then I need to use compose.https.yaml together with compose.proxy.yaml to get access to Frappe/ERPNext via https successfully. However documentation examples says that use compose.https.yaml exclusively is enough to get https working.

Fixed, thanks!

vrslev commented:
@kurtrottmann 

> My comment was related to mention this extra step more clearly in documentation.

@revant Can you update the docs?

revant commented:
> Should we update CONTRIBUTING.md before merge?
> Could you answer this one? Also we probably should update the docs.

I think I'll just merge this PR in the morning IST. We'll keep improving in smaller increments and break down into smaller issues.


3 areas to improve after the merge will be.

- docs
- contributing
- custom app how to.

we'll create these issues and focus there.

vrslev commented:
> We'll keep improving in smaller increments and break down into smaller issues.

Relieved to finally make proposed changes the upstream.

revant commented:
Godspeed

vrslev commented:
@revant Helm chart didn’t get released: https://github.com/frappe/frappe_docker/runs/5532914328?check_suite_focus=true
